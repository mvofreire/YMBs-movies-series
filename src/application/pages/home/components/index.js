export * from "./banner";
// export * from "./view-movies-series";
