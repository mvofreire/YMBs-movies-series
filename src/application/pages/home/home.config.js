import React from "react";
const HomePage = React.lazy(() => import("./Home"));

export default {
  exact: true,
  key: "home",
  label: "Home",
  path: "/home",
  component: HomePage
};
