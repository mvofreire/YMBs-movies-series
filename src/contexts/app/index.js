import React, { PureComponent } from "react";
import { getListMovies, getListSeries } from "services/movies-series";
import Storage from "utils/storage";
const AppContext = React.createContext();

export const TYPES = {
  MOVIE: "movie",
  SERIE: "serie"
};

export const VIEW_MODES = {
  LIST: "list",
  GRID: "grid"
};

export const normalizeFavorites = (list, favorites) =>
  list.map(item => ({ ...item, isFavorite: favorites.indexOf(item.id) > -1 }));

class AppContextProvider extends PureComponent {
  state = {
    viewType: TYPES.MOVIE,
    viewMode: VIEW_MODES.LIST,
    orderByRelease: false,
    isLoading: false,

    visibleDetail: false,
    itemDetail: {},

    data: [],
    favorites: []
  };

  setData = data => {
    this.setState({ data: normalizeFavorites(data, this.state.favorites) });
  };

  setFavorites = favorites => this.setState({ favorites });

  setType = (viewType, cb) => {
    this.setState({ viewType }, cb);
  };

  setOrder = (orderByRelease, cb) => {
    this.setState({ orderByRelease }, cb);
  };

  changeViewMode = mode => {
    this.setState({ viewMode: mode });
  };

  toggleFavorite = (id, cb = null) => {
    const { data, favorites } = this.state;
    const index = favorites.indexOf(id);

    let addFavorite = !(index > -1);
    if (!addFavorite) {
      favorites.splice(index, 1);
    } else {
      favorites.push(id);
    }

    const _data = normalizeFavorites(data, favorites);
    this.setState({ data: _data, favorites }, () => {
      Storage.set("favorites", favorites);
      cb && cb(addFavorite);
    });
  };

  toggleDetailFavorite = () => {
    const { itemDetail } = this.state;
    this.toggleFavorite(itemDetail.id, isFavorite => {
      this.setState({
        itemDetail: {
          ...itemDetail,
          isFavorite
        }
      });
    });
  };

  onClickItem = item => {
    this.setState({ itemDetail: item, visibleDetail: true });
  };

  closeDetails = () => {
    this.setState({ visibleDetail: false, itemDetail: {} });
  };

  render() {
    return (
      <AppContext.Provider
        value={{
          ...this.state,
          setData: this.setData,
          setFavorites: this.setFavorites,
          setType: this.setType,
          setOrder: this.setOrder,

          toggleDetailFavorite: this.toggleDetailFavorite,
          toggleFavorite: this.toggleFavorite,

          changeViewMode: this.changeViewMode,
          onClickItem: this.onClickItem,
          closeDetails: this.closeDetails
        }}
      >
        {this.props.children}
      </AppContext.Provider>
    );
  }
}

export { AppContext, AppContextProvider };
