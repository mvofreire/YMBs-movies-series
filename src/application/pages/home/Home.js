import React, { Fragment } from "react";
import { HomeBanner } from "./components";
import { ListEntertainment } from "components";
import { Divider } from "antd";
import { HomeContext, HomeContextProvider } from "contexts/home";

const Home = ({ data }) => (
  <HomeContextProvider>
    <HomeBanner />
    <Divider />
    <HomeContext.Consumer>
      {props => <ListEntertainment {...props} />}
    </HomeContext.Consumer>
  </HomeContextProvider>
);

export default Home;
