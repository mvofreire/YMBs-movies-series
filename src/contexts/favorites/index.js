import React, { PureComponent } from "react";
import { getFavoriteMovies, getFavoriteSeries } from "services/movies-series";
import Storage from "utils/storage";
import { AppContext, TYPES, VIEW_MODES } from "contexts/app";
import FavoritesManager from "utils/favorites";

const FavoritesContext = React.createContext();
class FavoritesContextProvider extends PureComponent {
  static contextType = AppContext;

  componentDidMount() {
    this.init();
    this.favoritesManager = new FavoritesManager();
  }

  init = () => {
    const { setType, changeViewMode } = this.context;

    changeViewMode(VIEW_MODES.LIST);
    setType(TYPES.MOVIE, this.loadData);
  };

  loadData = async () => {
    const { setData, setFavorites, orderByRelease } = this.context;
    const { viewType } = this.context;

    const favorites = this.favoritesManager.getList();
    const load =
      viewType === TYPES.MOVIE ? getFavoriteMovies : getFavoriteSeries;

    const data = await load(orderByRelease, favorites);

    setFavorites(favorites);
    setData(data);
  };

  changeType = type => {
    const { setType } = this.context;

    setType(type, this.loadData);
  };

  changeOrder = order => {
    const { setOrder } = this.context;

    setOrder(order, this.loadData);
  };

  removeFavorite = id => {
    const favorites = Storage.get("favorites") || [];
    const index = favorites.indexOf(id);
    favorites.splice(index, 1);
    Storage.set("favorites", favorites);

    this.loadData();
  };

  removeDetailFavorite = () => {
    const { itemDetail, closeDetails } = this.context;
    this.removeFavorite(itemDetail.id);
    closeDetails();
  };

  render() {
    const ctxValues = this.context;
    return (
      <FavoritesContext.Provider
        value={{
          ...ctxValues,
          changeType: this.changeType,
          changeOrder: this.changeOrder,
          toggleFavorite: this.removeFavorite,
          toggleDetailFavorite: this.removeDetailFavorite
        }}
      >
        {this.props.children}
      </FavoritesContext.Provider>
    );
  }
}

export { FavoritesContext, FavoritesContextProvider };
