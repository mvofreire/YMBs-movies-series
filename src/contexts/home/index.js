import React, { PureComponent } from "react";
import { getListMovies, getListSeries } from "services/movies-series";
import Storage from "utils/storage";
import { AppContext, TYPES, VIEW_MODES } from "contexts/app";

const HomeContext = React.createContext();
class HomeContextProvider extends PureComponent {
  static contextType = AppContext;

  componentDidMount() {
    this.init();
  }

  init = () => {
    const { setType, changeViewMode } = this.context;

    changeViewMode && changeViewMode(VIEW_MODES.LIST);
    setType && setType(TYPES.MOVIE, this.loadData);
  };

  loadData = async () => {
    const { setData, setFavorites, orderByRelease } = this.context;
    const { viewType } = this.context;

    const favorites = Storage.get("favorites") || [];
    const load = viewType === TYPES.MOVIE ? getListMovies : getListSeries;
    const data = await load(orderByRelease);

    setFavorites(favorites);
    setData(data);
  };

  changeType = type => {
    const { setType } = this.context;

    setType(type, this.loadData);
  };

  changeOrder = order => {
    const { setOrder } = this.context;

    setOrder(order, this.loadData);
  };

  render() {
    const ctxValues = this.context;
    return (
      <HomeContext.Provider
        value={{
          ...ctxValues,
          changeType: this.changeType,
          changeOrder: this.changeOrder
        }}
      >
        {this.props.children}
      </HomeContext.Provider>
    );
  }
}

export { HomeContext, HomeContextProvider };
