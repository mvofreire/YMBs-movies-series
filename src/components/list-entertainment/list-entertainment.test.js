import React from "react";
import { shape } from "prop-types";
import ReactDOM from "react-dom";
import { ListEntertainment } from ".";
import { shallow, mount } from "enzyme";
import { HomeContext, HomeContextProvider } from "contexts/home";

import ListItem from "./item-list";
import GridItem from "./item-grid";

describe("ListEntertainment", () => {
  const item = {
    id: "tt7366338",
    title: "Chernobyl",
    description: "",
    poster: "",
    year: "2019",
    released: "06 May 2019",
    genre: "Drama, History",
    director: "N/A",
    rating: "9.6",
    time: "330 min",
    type: "serie"
  };
  it("Wrapper must match snapshot ", () => {
    const wrapper = shallow(
      <HomeContextProvider>
        <HomeContext.Consumer>
          {props => <ListEntertainment {...props} />}
        </HomeContext.Consumer>
      </HomeContextProvider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("Should call callback on click title in ListItem", () => {
    const callback = jest.fn();
    const wrapper = mount(<ListItem item={item} onClickItem={callback} />);
    wrapper.find(".item-title").simulate("click");
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("Should call callback on click toggle Favorite", () => {
    const callback = jest.fn();
    const wrapper = mount(<ListItem item={item} toggleFavorite={callback} />);
    wrapper
      .find(".toggle-favorite")
      .first()
      .simulate("click");
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("Should call callback on click title in ListItem", () => {
    const callback = jest.fn();
    const wrapper = mount(<GridItem item={item} onClickItem={callback} />);
    wrapper.find(".item-title").simulate("click");
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("Should call callback on click toggle Favorite", () => {
    const callback = jest.fn();
    const wrapper = mount(<GridItem item={item} toggleFavorite={callback} />);
    wrapper
      .find(".toggle-favorite")
      .first()
      .simulate("click");
    expect(callback).toHaveBeenCalledTimes(1);
  });
});
