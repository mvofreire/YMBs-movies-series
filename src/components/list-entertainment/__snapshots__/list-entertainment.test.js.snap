// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListEntertainment Wrapper must match snapshot  1`] = `
<ContextProvider
  value={
    Object {
      "changeOrder": [Function],
      "changeType": [Function],
    }
  }
>
  <ContextConsumer>
    <Component />
  </ContextConsumer>
</ContextProvider>
`;
